tidy3d/plugins/invdes/postprocess.py

```diff
@@ -0,0 +1,131 @@
+# defines postprocessing classes for `InverseDesign` objects.
+from __future__ import annotations
+
+import abc
+import typing
+
+import autograd.numpy as anp
+import pydantic.v1 as pd
+
+import tidy3d as td
+
+from .base import InvdesBaseModel
+
+PostProcessFnType = typing.Callable[[td.SimulationData], float]
+
+
+class AbstractPostprocessOperation(InvdesBaseModel, abc.ABC):
+    """Abstract base class defining components that make up postprocessing classes."""
+
+    @abc.abstractmethod
+    def evaluate(self, sim_data: td.SimulationData) -> float:
+        """How to evaluate this operation on a ``SimulationData`` object."""
+
+
+class CustomPostprocessOperation(AbstractPostprocessOperation):
+    """A postprocessing operation to subclass and implement ones own ``evalute`` method for."""
+
+    def evaluate(self, sim_data: td.SimulationData) -> float:
+        """How to evaluate this operation on a ``SimulationData`` object."""
+        raise NotImplementedError("Must define the 'self.evaluate(sim_data) -> float' method.")
+
+    @classmethod
+    def from_function(cls, fn: PostProcessFnType, **kwargs) -> CustomPostprocessOperation:
+        """Create a ``CustomPostprocessOperation`` from a function of ``SimulationData``."""
+        cls.evaluate = lambda self, sim_data: fn(sim_data)
+        obj = cls(**kwargs)
+        return obj
+
+
+class ElementaryPostprocessOperation(AbstractPostprocessOperation, abc.ABC):
```

Comment:
Does this really need to inherit from `ABC` again?

Reply:
good question. I'm starting to doubt how I thought ABC classes worked. The idea though was that this `ElementaryPostprocessOperation` is still abstract in the sense that we just use it for grouping / subclassing. We don't want the user to initialize one.

I thought that meant making it inherit directly from `abc.ABC` but not I don't know. seems you can still initlize it. Have I been using `ABC` wrong this whole time? when should we use it?

Reply:
So in principle this works:
```python
class A(abc.ABC):
    @abc.abstractmethod
    def f(self):
        ...

class B(A):
    def g(self):
        ...

b = B()  # TypeError
```
However, if you use composition like, e.g., `A(float, abc.ABC)`, then `A` can be instantiated and is not abstract anymore. So yeah, since everything inherits from `BaseModel`, I don't think any of our `ABC`s really work as intended :smile:
This can be solved through some fiddling with `abc.ABCMeta` but I have to double check, forgot how exactly that works.